src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -810,6 +840,32 @@ - (void)_handleReportBegin
     }
 }
 
+// assume lock is held
+- (NSDictionary<NSNumber *, NSDictionary *> *)_attributesForCluster:(MTRClusterPath *)clusterPath
+{
+    NSMutableDictionary * attributesToReturn = [NSMutableDictionary dictionary];
+    for (MTRAttributePath * attributePath in _readCache) {
+        if ([attributePath.endpoint isEqualToNumber:clusterPath.endpoint] && [attributePath.cluster isEqualToNumber:clusterPath.cluster]) {
+            attributesToReturn[attributePath.attribute] = _readCache[attributePath];
+        }
+    }
+    return attributesToReturn;
+}
+
+// assume lock is held
+- (NSDictionary<MTRClusterPath *, MTRDeviceClusterData *> *)_clusterDataForPaths:(NSSet<MTRClusterPath *> *)clusterPaths
+{
+    NSMutableDictionary * clusterDataToReturn = [NSMutableDictionary dictionary];
+    for (MTRClusterPath * clusterPath in clusterPaths) {
+        NSNumber * dataVersion = _clusterData[clusterPath].dataVersion;
```

Comment:
Could you clarify what you mean?

Reply:
Assuming you were asking about if both data version and attributes are nil - I've made a change here to account for that in the latest commit.

Reply:
What happens if `_clusterData[clusterPath]` is nil?  Or can that not happen?

Reply:
That should not happen as far as I understand. We have a fast enumeration on the dictionary with the keys in `clusterPath` and so the dictionary should definitely have values for it.

Modification of dictionary during enumeration is covered by the objc runtime.

And moreover, even if it's nil, the new nil checks I have should cover it as well, because nil objects in objc would return nil properties.